m_etcd/integration_test.go

```diff
@@ -12,6 +12,115 @@ import (
 	"github.com/lytics/metafora/statemachine"
 )
 
+// TestSleepTest is an integration test for all of m_etcd's components.
+//
+func TestSleepTest(t *testing.T) {
+	etcdc, hosts := testutil.NewEtcdClient(t)
+	t.Parallel()
+	const namespace = "sleeptest-metafora"
+	const sleepingtasks = "sleeping-task1"
+
+	const recursive = true
+	etcdc.Delete(namespace, recursive)
+
+	holdtask := make(chan bool)
+	h := func(tid string, cmds <-chan statemachine.Message) statemachine.Message {
+
+		if tid == sleepingtasks {
+			sleeptil := 5 * time.Second
+			nextstarttime := (time.Now().Add(sleeptil))
+			t.Logf("sleeping task:%v sleepfor:%v", tid, nextstarttime)
+			<-holdtask
+			return statemachine.Message{Code: statemachine.Sleep, Until: &nextstarttime}
+		}
+
+		cmd := <-cmds
+		t.Logf("non sleeping task:%v", tid)
+
+		return cmd
+	}
+
+	newC := func(name, ns string) *metafora.Consumer {
+		coord, hf, bal, err := m_etcd.New(name, ns, hosts, h)
+		if err != nil {
+			t.Fatalf("Error creating new etcd stack: %v", err)
+		}
+		cons, err := metafora.NewConsumer(coord, hf, bal)
+		if err != nil {
+			t.Fatalf("Error creating consumer %s:%s: %v", ns, name, err)
+		}
+		go func() {
+			cons.Run()
+			t.Logf("Consumer:%s exited.", name)
+		}()
+		return cons
+	}
+
+	assertRunning := func(tid string, cons ...*metafora.Consumer) {
+		found := false
+		for _, c := range cons {
+			tasks := c.Tasks()
+			if len(tasks) > 0 && found {
+				t.Fatal("Task already found running but another task is running on a different consumer")
+			}
+			if len(tasks) > 1 {
+				t.Fatalf("Expected at most 1 task, but found: %d", len(tasks))
+			}
+			if len(tasks) == 1 && tasks[0].ID() == tid {
+				found = true
+			}
+		}
+		if !found {
+			t.Fatalf("Could not find task=%q", tid)
+		}
+	}
+
+	// Start 2 consumers
+	cons1 := newC("node1", namespace)
+	cons2 := newC("node2", namespace)
+
+	// Create clients and start some tests
+	cliA := m_etcd.NewClient(namespace, hosts)
+
+	if err := cliA.SubmitTask(sleepingtasks); err != nil {
+		t.Fatalf("Error submitting task1 to a: %v", err)
+	}
+
+	// Give consumers a bit to pick up tasks
+	time.Sleep(500 * time.Millisecond)
+
+	assertRunning(sleepingtasks, cons1, cons2)
+
+	holdtask <- true
+	// Give consumers a bit to pick up tasks
+	time.Sleep(500 * time.Millisecond)
+
+	assertRunning(sleepingtasks, cons1, cons2) // not sure if this should be true or false.
+
+	wait1 := make(chan bool)
+	go func() {
+		defer close(wait1)
+		// Shutdown
+		cons1.Shutdown()
+		cons2.Shutdown()
+	}()
+
+	timeout := time.NewTimer(5 * time.Second)
```

Comment:
Hm, I think this test could be improved by decreasing this timeout to be <5s. Here's my thinking:

* T1 = test start
* T1 < task starts (TS) < T2
* T2 = T1 + 500ms = task is running (line 92)
* T3 = task start + 5s = when the task should wake up
* T4 = T2 + 500ms + holdchan latency = when sleep message w/T3 timestamp is returned to the statemachine (line 96)
* T5 < (T4 + 5s deadline) = T5 is the time when the shutdowns complete

Since `T4 > TS` then `T5 > T3`. This means the task wakes up *before* the shutdown deadline is reached.

What we want to test is that a *sleeping* task shuts down cleanly. This tests that a sleeping task shuts down *upon waking up.*


At least I think :) 

**I'm going to pull your branch and play around**